Adjust fragment radius according to ablation

gsc10 opened the pull request:
Unless the chain reaction cloud model is on, fragments do not change radius when mass is lost by ablation. I think this should be fixed to adjust radius (and hence area) according to mass loss for solid fragments (not clouds).

gsc10 committed:
Adjust fragment radius according to ablation

Unless the chain reaction cloud model is on, fragments do not change radius when mass is lost by ablation. I think this should be fixed to adjust radius (and hence area) according to mass loss for solid fragments (not clouds).

acse-ds2419 commented:
I remember that I was puzzled back during ACSE-4 that before break-up, we assumed that the radius perpendicular to the trajectory would remain constant. I think we said the idea is that the meteoroid only shrinks along one axis, therefore becoming shaped like an ellipsoid. As I showed in my report, Avramenko's formula for dr/dt makes sure that the fragment in our model keeps its spherical shape as it ablates.

Do you think that other papers just keep r constant (before break-up) for convenience sake?

acse-ds2419 commented:
I'm wondering whether we should make this "shrink r when ablation happens" an option/model parameter, or whether the model should always do that. But I'm all for decoupling it from the choice of debris cloud model.

gsc10 commented:
Keeping r constant is not a big issue for large meteoroids, so the choice of assumption in previous work probably depended on the size of objects under consideration and whether or not ablation was considered/important.

For the application of small craters on Mars, this assumption causes big problems because the fragments lose most of their mass, but (currently) don't change radius.

If we make it an optional parameter, I think the default should be that r does change with ablation.

acse-ds2419 commented on src/fcm.cpp:
```diff
@@ -57,7 +57,8 @@ offset fcm::dfdt(const Fragment& fragment, const FCM_params& params, const FCM_s
                                         params.cloud_disp_coeff, fragment.density());
         } else {
-            result.dr = result.d2r = 0;
+	    result.d2r = 0;
+	    result.dr = drdt_ablation(fragment.radius(), fragment.mass(), result.dm);
```
This now means that the if clause does exactly the same thing as the else clause. I like using `drdt_ablation()` here instead of `drdt_chainreaction`.

⇨ I let's make it a parameter in the of the FCM_settings struct whether to use dr/dt = `drdt_ablation()` or to use dr/dt = 0. Then the equivalent class in python must be modified. I'll make that change tomorrow.

acse-ds2419 commented on include/ode.hpp:
```diff
@@ -591,6 +591,19 @@ constexpr auto drdt_chainreaction(double rp, double strength, double r, double m
+ * @param dmdt : ablation mass loss, in [kg/s]
+ */
+constexpr auto drdt_ablation(double r, double m, double dmdt) noexcept {
+    return r/m * dmdt/3;
```
Re-use this function in `drdt_chainreaction()` -> then we don't have code duplication

acse-ds2419 reviewed (changes requested):
Whether or not to use dr/dt = `drdt_ablation()` should be a model parameter that can be set in python. Should default to true.

acse-ds2419 committed:
added folder for python virtual env to gitignore

acse-ds2419 committed:
calling drdt_ablation() from drdt_chainreaction()
to avoid code duplication

acse-ds2419 committed:
- added parameter "meteoroid_const_r" to
   AblationModel
- changed dfdt to switch between the
   chain reaction model and the constant r
   model based on the AblationModel param
- changed interface of FCM_settings constructor
   and set defaults for flat_planet and fixed_timestep

acse-ds2419 committed:
exposing and propagating additional parameter
for ablation model to the python side

acse-ds2419 committed:
added test case for new ablation model parameter

acse-ds2419 committed:
Merge branch 'master' into gsc-fix-ablation

acse-ds2419 commented:
Hi @gsc10 I finally had some time for the changes that I described. Please have a look and let me know what you think.

Unfortunately, after merging with master we've inherited a failing cpp test (`fcm_tests.cpp::dEdz/it_works), which I guess comes from the change in the 'fix projection issue' PR. I'll try solving that next week.

gsc10 reviewed (commented):
Did you mean to knock out `test_dedz()` and `test_fragmentation()` ? Otherwise, looks good!

acse-ds2419 committed:
marked ODE_tests::holsapple_regolith and
ODE_tests::holsapple_granular_soil as XFAIL

acse-ds2419 commented:
> 
> 
> Did you mean to knock out `test_dedz()` and `test_fragmentation()` ? Otherwise, looks good!

I marked the two failing tests in ode_tests as XFAIL, since this is more of a theoretical check. I left a comment about why it should be passing, and that it needs investigation. But these two have been failing ever since I implemented them, so it doesn't indicate a problem with a PR if they fail.

As for `fcm_tests.cpp::dEdz/it_works`, this one's concerning and needs some investigation.

acse-ds2419 committed:
convergence_tests.py: only importing matplotlib
if run as __main__ -> makes pytest mode faster

acse-ds2419 committed:
fixed error in cpp unit test

acse-ds2419 committed:
moved importing matplotlib to plot_craters, since
this is the only function using it.
-> it is not part of our unit tests, i.e. only importing
    it here speeds these up

acse-ds2419 committed:
skipping the mars atmosphere API test

acse-ds2419 commented:
Fixed the remaining failing test :) It was simply a logic error that I made back when I designed the test
